File: elixir/tournament/lib/tournament.ex
defmodule Tournament do
  # NOTE: One has to "require Tournament" in iex> to make use of the macros (i.e. Tournament.match_result/0, 1 or 2) defined by "defrecord" there
  require Record
  Record.defrecord(:match_result, team: "Name", result: "loss", points: 0)
  @type match_result :: record(:match_result, team: String.t(), result: String.t(), points: non_neg_integer())

  defmodule TeamStats do
    defstruct [:name, :played, :win, :draw, :loss, :points]
    @type t :: %__MODULE__{name: String.t(), played: pos_integer(), win: non_neg_integer(), draw: non_neg_integer(), loss: non_neg_integer(), points: non_neg_integer()}
  end
  @doc """
  Given `input` lines representing two teams and whether the first of them won,
  lost, or reached a draw, separated by semicolons, calculate the statistics
  for each team's number of games played, won, drawn, lost, and total points
  for the season, and return a nicely-formatted string table.

  A win earns a team 3 points, a draw earns 1 point, and a loss earns nothing.

  Order the outcome by most total points for the season, and settle ties by
  listing the teams in alphabetical order.
  """
  @spec tally(input :: list(String.t())) :: String.t()
  def tally(input) do
    header = "Team                           | MP |  W |  D |  L |  P"
    teams =
      input
      |> Enum.map(&String.split(&1, ";"))
      |> Enum.filter(&(Enum.count(&1) == 3))
      |> Enum.flat_map(fn [team_a, team_b, result] ->
        case result do
          "win"  -> [match_result(team: team_a, result: result, points: 3), match_result(team: team_b, result: "loss", points: 0)]
          "draw" -> [match_result(team: team_a, result: result, points: 1), match_result(team: team_b, result: result, points: 1)]
          "loss" -> [match_result(team: team_a, result: result, points: 0), match_result(team: team_b, result: "win",  points: 3)]
          _      -> []
        end
      end)
      |> Enum.group_by(fn rec -> match_result(rec, :team) end)
      |> Enum.map(fn {team, results} -> %TeamStats{
        name: team,
        played: results |> Enum.count,
        win: results |> Enum.filter(&(match_result(&1, :result) == "win")) |> Enum.count,
        draw: results |> Enum.filter(&(match_result(&1, :result) == "draw")) |> Enum.count,
        loss: results |> Enum.filter(&(match_result(&1, :result) == "loss")) |> Enum.count,
        points: results |> Enum.map(&match_result(&1, :points)) |> Enum.sum }
      end)
      |> Enum.sort_by(fn stats -> stats.name end)
      |> Enum.sort_by(fn stats -> stats.points end, &>=/2)
      |> Enum.map(fn stats ->
        fmt_name = stats.name |> String.pad_trailing(30)
        fmt_played = stats.played |> to_string() |> String.pad_leading(2)
        fmt_win = stats.win |> to_string() |> String.pad_leading(2)
        fmt_draw = stats.draw |> to_string() |> String.pad_leading(2)
        fmt_loss = stats.loss |> to_string() |> String.pad_leading(2)
        fmt_points = stats.points |> to_string() |> String.pad_leading(2)
        "#{fmt_name} | #{fmt_played} | #{fmt_win} | #{fmt_draw} | #{fmt_loss} | #{fmt_points}"
      end)
    [header | teams] |> Enum.join("\n")
  end
end

File: fsharp/tournament/Tournament.fs
module Tournament
    
let tally (input: string list): string list = 
    let header = $"Team                           | MP |  W |  D |  L |  P"
    let tl = 
        input 
        |> Seq.map (fun s -> s.Split([|';'|], System.StringSplitOptions.RemoveEmptyEntries))  // split each line
        |> Seq.filter (fun arr -> arr.Length >= 3)  // get rid of badly formatted results or empty rows
        |> Seq.map (fun arr -> {| teamA = arr.[0]; teamB = arr.[1]; result = arr.[2] |} ) // restructure the matches by named fields
        |> Seq.collect (fun m ->   // now split every match into two anonymous result records for each of the teams
            if m.result = "draw" then 
                seq [ {| team = m.teamA; result = 'D'; points = 1 |}; {| team = m.teamB; result = 'D'; points = 1 |} ]
            else // add some logic to correctly record the winning or losing team
                seq [ {| team = (if m.result = "win" then m.teamA else m.teamB); result = 'W'; points = 3 |}; 
                {| team = (if m.result = "win" then m.teamB else m.teamA); result = 'L'; points = 0 |} ] )
        |> Seq.groupBy (fun r -> r.team)  // group these results by team
        |> Seq.map (fun g ->  // get the necessary statistics from the results sequence of each team
            let results = snd g
            {|  Name = fst g
                MP = results |> Seq.length
                W = results |> Seq.filter (fun r -> r.result = 'W') |> Seq.length
                D = results |> Seq.filter (fun r -> r.result = 'D') |> Seq.length
                L = results |> Seq.filter (fun r -> r.result = 'L') |> Seq.length
                Points = results |> Seq.sumBy (fun r -> r.points)
                |} )
        |> Seq.sortBy (fun r -> r.Name) // sort first by name
        |> Seq.sortByDescending (fun r -> r.Points) // and then by points, this leaves the correct alphabetical order for teams with equal points
        |> Seq.map (fun r -> $"{r.Name,-30} | {r.MP,2} | {r.W,2} | {r.D,2} | {r.L,2} | {r.Points,2}")
        |> Seq.toList
    header :: tl // add the header to the front